src/lib.rs

```diff
@@ -173,164 +194,388 @@ impl<const OSC_RAW_BUF_SIZE: usize> Parser<OSC_RAW_BUF_SIZE> {
+                self.state = State::Ground
             },
-            Action::Collect => {
-                if self.intermediate_idx == MAX_INTERMEDIATES {
-                    self.ignoring = true;
-                } else {
-                    self.intermediates[self.intermediate_idx] = byte;
-                    self.intermediate_idx += 1;
-                }
+            0x1B => {
+                self.reset_params();
+                self.state = State::Escape
             },
-            Action::Param => {
-                if self.params.is_full() {
-                    self.ignoring = true;
-                    return;
-                }
+            _ => (),
+        }
+    }
 
-                match byte {
-                    b';' => {
-                        self.params.push(self.param);
-                        self.param = 0;
-                    },
-                    b':' => {
-                        self.params.extend(self.param);
-                        self.param = 0;
-                    },
-                    _ => {
-                        // Continue collecting bytes into param
-                        self.param = self.param.saturating_mul(10);
-                        self.param = self.param.saturating_add((byte - b'0') as u16);
-                    },
-                }
+    #[inline]
+    fn action_csi_dispatch<P: Perform>(&mut self, performer: &mut P, byte: u8) {
+        if self.params.is_full() {
+            self.ignoring = true;
+        } else {
+            self.params.push(self.param);
+        }
+        performer.csi_dispatch(self.params(), self.intermediates(), self.ignoring, byte as char);
+
+        self.state = State::Ground
+    }
+
+    #[inline]
+    fn action_hook<P: Perform>(&mut self, performer: &mut P, byte: u8) {
+        if self.params.is_full() {
+            self.ignoring = true;
+        } else {
+            self.params.push(self.param);
+        }
+        performer.hook(self.params(), self.intermediates(), self.ignoring, byte as char);
+        self.state = State::DcsPassthrough;
+    }
+
+    #[inline]
+    fn action_collect(&mut self, byte: u8) {
+        if self.intermediate_idx == MAX_INTERMEDIATES {
+            self.ignoring = true;
+        } else {
+            self.intermediates[self.intermediate_idx] = byte;
+            self.intermediate_idx += 1;
+        }
+    }
+
+    // When we transition from the 0x3A.
+    #[inline]
+    fn action_subparam(&mut self) {
+        if self.params.is_full() {
+            self.ignoring = true;
+        } else {
+            self.params.extend(self.param);
+            self.param = 0;
+        }
+    }
+
+    // When we transition from the 0x3B.
+    #[inline]
+    fn action_param(&mut self) {
+        if self.params.is_full() {
+            self.ignoring = true;
+        } else {
+            self.params.push(self.param);
+            self.param = 0;
+        }
+    }
+
+    // Neither 0x3A nor 0x3B.
```

Comment:
I think it's better to document what the function does, rather than how it is entered.

In this case `paramnext` advances while inside a parameter, without terminating that parameter.